user/src/com/google/gwt/user/client/ui/RootPanel.java

```diff
@@ -258,15 +253,6 @@ private static native Element getRootElement() /*-{
     return $doc;
   }-*/;
 
-  private static void hookWindowClosing() {
```

Comment:
I think we should somehow keep that logic. Introducing `pagehide` event first and then use it here could be a solution. It fires unreliably on mobile, just like `beforeunload`, but it is fired after `beforeunload` so we can be sure the user hasn't canceled unloading the page.

If we fully delete this logic here, then `RootPanel.detachOnWindowClose()` is now a pretty bad method name and all Widgets created using `wrap(Element)` methods (also custom widgets) do not receive calls to `Widget.onDetach` anymore. As a result some code in end user applications isn't executed anymore.

Reply:
Since it is unreliable, is it important to keep? The goal according to javadoc for detachNow and detachOnWindowClose is to ensure that objects are consistently GC'd, which is already certain to be true (since IE9 and FF 1.6 are long dead).

I agree that for (inconsistent) consistency we _could_ replace this with `pagehide` - but `pagehide` comes with different semantics that may confused application authors who previously assumed this behavior. For example, if using the `bfcache`, the app may continue to function after `pagehide`, and each of those widgets that was removed will need to be re-attached to resume handling events.  That's a pivot to a much bigger feature than what RootPanel is trying to support - and it may not be consistent with the current API, right?

Reply:
Hmm yeah it would be unfortunate if `bfcache` restores the page from memory and nothing is in attached state anymore. The code could check `pagehide.persisted` and only detach if it is `false`. But who knows how often it will really be `false`.

My primary thinking was that we don't know what people really do in `onDetach` or as a consequence in `onUnload`, e.g. `postMessage` or something like that. Calling it unreliably might be better than not calling it at all. But I don't really have a strong opinion about it. However if we remove it here, then `RootPanel.detachOnWindowClose()` should probably be marked deprecated for removal with updated JavaDoc and `RootPanel.addToDetachList()` should be added as a replacement.

Reply:
I'm really loathe to expand the API here, but would instead prefer to focus efforts on the updated modules and enhancing them - no jsni, no generators, no rebind rules, and slowly "strangle" the old gwt-user code until it is phased out.

Reply:
On that note, I guess that either detachWidgets() should be removed, or made public and deprecated so that users can still call into it if they really want the old behavior? I've put up a patch doing this, with some documentation on alternatives.